fix(ast, codegen, linter): panics in fixers.

rzvxa opened the pull request:
Closes #5434

https://github.com/oxc-project/oxlint-ecosystem-ci/actions/runs/10685877915/job/29619941099

rzvxa commented:
* **#5431** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5431?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rzvxa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

camc314 commented:
does this fix all of the panics in that action?

Boshen reviewed (changes requested):
Oh wait I forgot to say that we should add some tests. I intend to change how codegen deal with source_text at some point.

Boshen commented on crates/oxc_ast/src/ast_impl/literal.rs:
how do I block the merge queue lol

Boshen reviewed (dismissed):
block!

rzvxa commented:
> Oh wait I forgot to say that we should add some tests. I intend to change how codegen deals with source_text at some point.

I am converting this to draft while I'm trying to find a good way to test it.

Do you have any specific approach for testing this in mind? What part should I test? The reason behind the panic was that I assumed `source_text` in the codegen would always be set similarly to formatter, That's why we encountered out-of-bound issues when trying to generate regexes. Now that fixers are operating on a codegen with source_text, This error is resolved.

The `pattern.source_text` method is a dead simple function and `to_string`/`Display` has its own set of tests.

We use source_text when available because the regex pattern loses some of the information from the original input so subsequently in some cases, we can't recreate the exact original input with the `Display` trait. We might even want to do the same thing for regex flags because as of now we change their order(However this is a small issue compared to the pattern itself which is more annoying if it changes unintentionally).

Boshen commented:
The linter rule tests should have at least 1 fixer test.

We should also have one integration test in apps/oxlint/src/lint/mod.rs with the fixture in apps/oxlint/fixtures/linter/debugger.js

rzvxa commented on crates/oxc_codegen/tests/integration/unit.rs:
```diff
@@ -63,6 +63,33 @@ fn unicode_escape() {
 }
 
+#[test]
+fn regex() {
```
Added some tests here, It would've caught the issue we had.
I'm not sure exactly how to write the oxlint integration test, Maybe we can write a temp file and apply a bunch of fixes to it to make sure the whole lint cycle is working as expected but I feel that should be its own PR.

rzvxa replied:
There are also tests available for rules with fixers, The only reason that it was uncaught was that none of our regex fixers were using codegen, The root of the issue was that codegen was panicking when the source wasn't available to it.
These tests make sure we generate the correct regexes with or without the source. Now having the original source in the linter becomes a decision rather than a bug and that's why I have difficulties imagining a useful unit test for it.

rzvxa committed:
fix(ast, codegen, linter): panics in fixers. (#5431)

Closes #5434

https://github.com/oxc-project/oxlint-ecosystem-ci/actions/runs/10685877915/job/29619941099
